### PR TITLE
Implement rule sets and rule priority

### LINKED ISSUE #193: [Discussion] Proposal for a rule sets implementation
> Note: I may be misunderstanding how distributed slices and macros work in Rust. These are just some thoughts that I have! Putting it out there for discussion :)

# Overview

Currently, we keep all rules in a distributed slice, so there is no way to enable or disable rules for specific solvers.
When we apply the `rewrite` function, all existing rules get applied.

I believe we are currently planning to solve this using cargo features (essentially, conditionally building rules based on which features are turned on). However, this approach would not allow the user to define their own separate sets of rules (aka rule classes or rule kinds).

# Proposal

My idea broadly is to:
- Create a struct `RuleSet` that would hold:
    - `rules` - A vector of pairs consisting of a reference to a `Rule` and a priority of this rule in this specific ruleset
    - `dependencies` - Its dependencies (vector of references to other `RuleSet`s that need to be enabled for it to work) 
    - `antidependencies` - Its anti-dependencies (`RuleSet`s that conflict with it)
    - `priority` - The priority of the ruleset
- Create a separate distributed slice that would hold these `RuleSet`s, a function to register a ruleset, and a function to get a `RuleSet` by its name (similar to `get_rule_by_name`)
- Add an argument to our `register_rule` macro - a vector of references to `RuleSet` objects. The macro should automatically add a reference to this rule to all of them
- Pass a list of strings representing ruleset names to the `rewrite` function and make it return a `Result<Expression, Error>` - either a rewritten expression, or an error if ruleset names are not found or if conflicting rulesets are given
- Before starting the rewrite, the `rewrite function` should:
    - Fetch all the `RuleSet` objects from the slice by their names (that were passed to it as arguments)
    - Recursively resolve their dependencies and make a final `Set` of `RuleSet`'s (or error if any of them conflict)
    - Resolve the final list of rules and priorities (see pseudocode bellow)
    - Rewrite the AST by applying these rules

# The `RuleSet` struct

```rust
struct RuleSet {
    priority: i32,
    rules: Vec<(&Rule, i32)>,
    dependencies: Vec<&RuleSet>,
    antidependencies: Vec<&RuleSet>,
}
```

# Resolving rulesets (pseudocode)

```python
function resolve_rule_sets(names)
    rule_sets: Vec<RuleSet>

    for name in names:
        rule_set = getRuleSetByName(name)
        if error, return error

        dependencies = resolve_dependencies(rule_set)
        if error, return error

        for new_rs in dependencies:
            check conflicts in rule_sets with new_rs
            if conflict, return error

            add new_rs to rule_sets

    return rule_sets
```

# Resolving the final set of rules (pseudocode)

```python
function resolve_rules(rule_sets)
  rules: Vec<(rule, priority, from_rule_set)>

  for rule_set in rule_sets:
    for rule, priority in rule_set.rules:
      if rule not in rules:
        add rule to rules
      else:
        if rule_set.priority > rule.from_rule_set.priority:
          update priority of rule
  
  return rules   # We can discard the reference to from_rule_set, or keep this information if it's useful
```

# Rewriting
```rust
pub fn rewrite(expression: &Expression, rule_set_names: Vec<string>) -> Result<Expression, Error> {
    //  (removed) let rules = get_rules();
    let rule_sets: Vec<&RuleSet> = resolve_rule_sets(rule_set_names);
    let rules: Vec<(&Rule, priority, &RuleSet)> = resolve_rules(rule_sets);
    // Rest of our code is the same
}
```

# New `register_rule` macro
```rust
#[register_rule([&minionRS, &baseRS], 5)]
fn my_rule(expr: &Expr) -> Result<Expr, RuleApplicationError> { ... }
```

Would not only add the rule to the rule table, but also add it to the given rule sets with the given priority:

```
minionRS.rules.push([MyRule, 5]);
baseRS.rules.push([MyRule, 5]);
```

# Why? 
- This gives us a flexible way to define rule sets and rule priorities
- Also removes the need to have any metadata in the rule itself
- Theoretically, the user may define as many rule set as they like (and these rulesets may depend on other custom rulesets or on any of our base rulesets)
- Also gives us a way to mark some rule sets as conflicting (i.e. the rules for converting to conjunctive normal form vs disjunctive normal form would conflict to each other, so marking their rulesets as anti-dependencies of each other would prevent the rule engine from applying both and looping)

# Why not?
To be honest, I'm not sure this is possible in plain Rust

### LINKED ISSUE #224: Ensure that rules have deterministic ordering
As per #223 , our rewrite system should be confluent / deterministic. However, the ordering of rules in the registry is currently undefined behaviour, and could change from compiler to compiler and on different versions of linkme (https://docs.rs/linkme/).

We should ensure that rules have a set ordering.

**Possible solution**

We currently assemble the rules "distributed slice" into a vector on every call to `conjure_rules::get_rules()`. We could instead use a [OnceCell/OnceLock](https://doc.rust-lang.org/std/sync/struct.OnceLock.html) to initialise this vector the first time `get_rules` is called and use the same one for successive calls. The constructor for this vector would impose ordering (such as lexical ordering).

### PR DESCRIPTION
**Pull request by gskorokhod**
Should close #193 by implementing rule sets and rule priorities as discussed.
As a consequence of that, also closes #224 (since we will now always have a vector of rules sorted by priority and name)

What's left:

- [x] Convert all rules to use new `register_rule` macro
- [x] Actually create and populate rule sets
- [x] Clean up and document code
- [x] Add wiki entries
- [x] Make sure it compiles and all tests still pass

### CODE REVIEW

**Comment by gskorokhod** [comment]
All seems to be working now!

File: conjure_oxide/src/rules/base.rs [hunk 1]
```diff
@@ -1,22 +1,25 @@
...
+ * Remove nothing's from expressions:
  * ```text
  * and([a, nothing, b]) = and([a, b])
  * sum([a, nothing, b]) = sum([a, b])
  * sum_leq([a, nothing, b], c) = sum_leq([a, b], c)
  * ...
  * ```
 */
-#[register_rule]
+#[register_rule(("Base", 50))] // ToDo: I am not sure what the priority should be - discuss
```
Comment (niklasdewally) [inline conjure_oxide/src/rules/base.rs:22]: Is this resolved?

Maybe some "magic numbers" might be helpful in the crate-level documentation?

File: conjure_oxide/src/rules/cnf.rs [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
+
 /**
 * Distribute `not` over `and` (De Morgan's Law):
 
 * ```text
 * not(and(a, b)) = or(not a, not b)
 * ```
  */
-#[register_rule]
+#[register_rule(("CNF", 20))] // ToDo: not sure about the priority - discuss
```
Comment (niklasdewally) [inline conjure_oxide/src/rules/cnf.rs:18]: Same as above

File: conjure_oxide/tests/generated_tests.rs [hunk 3]
```diff
@@ -83,7 +85,21 @@ fn integration_test(path: &str, essence_base: &str) -> Result<(), Box<dyn Error>
 
     // ToDo: There is A LOT of duplication here. We should refactor this to a function. In my defense, it was midnight when I wrote this.
```
Comment (niklasdewally) [inline conjure_oxide/tests/generated_tests.rs:86]: Still todo?

**Review by niklasdewally — APPROVED** [review]
On the whole, LGTM!

To be pedantic, what is the difference between the conjure_rules and conjure_rule_sets crate? I don't quite get the division - especially as the rulesets and rule functions kind of overlap, and rule-sets uses conjure_rules_proc_macro under the hood anyways?

**Comment by gskorokhod** [comment]
> On the whole, LGTM!
> 
> 
> 
> To be pedantic, what is the difference between the conjure_rules and conjure_rule_sets crate? I don't quite get the division - especially as the rulesets and rule functions kind of overlap, and rule-sets uses conjure_rules_proc_macro under the hood anyways?

That's fair, I wasn't sure whether to separate them or now. Should I put them all in conjure_rules?

**Comment by niklasdewally** [comment]
> > On the whole, LGTM!
> > To be pedantic, what is the difference between the conjure_rules and conjure_rule_sets crate? I don't quite get the division - especially as the rulesets and rule functions kind of overlap, and rule-sets uses conjure_rules_proc_macro under the hood anyways?
> 
> That's fair, I wasn't sure whether to separate them or now. Should I put them all in conjure_rules?

I think so - its all part of the same system imo.

(Consider) Is there any reason why the docs are on the wiki? Could this be put as crate documentation using rustdoc instead? (If the text is too long to fit in the lib.rs file, it is possible to include a markdown file inside rust doc comments)

**Comment by ozgurakgun** [comment]
Ooops @gskorokhod I tried to merge main onto this PR but a test fails now.

**Comment by niklasdewally** [comment]
> @ozgurakgun huh, it's only one of them as well. I'll have a look!

Was worried i broke something, but looks like rule application ordering has just changed  :)

**Comment by gskorokhod** [comment]
It's a pretty interesting error!

From what I understand, sometimes, get_rules() gets called by multiple threads, and they both try to initialise the hashmap at the same time, which errors because OnceLock just doesn't let the second thread get a lock:
image

**Comment by niklasdewally** [comment]
> It's a pretty interesting error!
> 
> From what I understand, sometimes, get_rules() gets called by multiple threads, and they both try to initialise the hashmap at the same time, which errors because OnceLock just doesn't let the second thread get a lock:
> image
> 

Yes, cargo tests run in threaded, so we should be thread safe!

**Comment by niklasdewally** [comment]
@gskorokhod it's interesting that oncelocks don't block...

Maybe a oncecell in a rwlock might be more what you want here?

**Comment by gskorokhod** [comment]
> @gskorokhod it's interesting that oncelocks don't block...
> 
> Maybe a oncecell in a rwlock might be more what you want here?

although this would probably be cleaner?

**Comment by niklasdewally** [comment]
> > @gskorokhod it's interesting that oncelocks don't block...
> > 
> > Maybe a oncecell in a rwlock might be more what you want here?
> 
> although this would probably be cleaner?

My understanding is you want to do:

1. Read the cell
2. If it's blank, init it.
3. Read again


Where many threads can read the initialised cell, but no thread can read while another thread is initialising the oncecell.


Rwlock does the "no reading while writing" bit, oncecell does the "this may not exist" bit

Is that what you want here?

**Comment by gskorokhod** [comment]
Huh

image

When set() returns something, it is in fact a value that's been put in the cell (by another thread), so instead of panicking we may just return it 

(it will allways be the same anyway because it's known at compile time and doesnt change)

**Comment by gskorokhod** [comment]
> > > @gskorokhod it's interesting that oncelocks don't block...
> > > Maybe a oncecell in a rwlock might be more what you want here?
> > 
> > 
> > although this would probably be cleaner?
> 
> My understanding is you want to do:
> 
> 1. Read the cell
> 2. If it's blank, init it.
> 3. Read again
> 
> Where many threads can read the initialised cell, but no thread can read while another thread is initialising the oncecell.
> 
> Rwlock does the "no reading while writing" bit, oncecell does the "this may not exist" bit
> 
> Is that what you want here?

Yes, but I think it's not necessary. I just didn't RTFM of OnceLock enough so wasn't sure whether it blocks (turns out it does, but in a slightly different way)

File: conjure_oxide/src/rule_engine/resolve_rules.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
+use conjure_core::rule::Rule;
+use conjure_rules::get_rule_set_by_name;
+use conjure_rules::rule_set::RuleSet;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Display;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum ResolveRulesError {
```
Comment (niklasdewally) [inline conjure_oxide/src/rule_engine/resolve_rules.rs:9]: (Pedantic) Make this #[non_exhaustive] ?

File: conjure_oxide/src/rule_engine/resolve_rules.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        match self {
+            ResolveRulesError::RuleSetNotFound => write!(f, "Rule set not found."),
+        }
+    }
+}
+
+/**
+* Helper function to get a rule set by name, or return an error if it doesn't exist.
+* @param `rule_set_name` The name of the rule set to get.
+* @returns The rule set with the given name or RuleSetError::RuleSetNotFound if it doesn't exist.
```
Comment (niklasdewally) [inline conjure_oxide/src/rule_engine/resolve_rules.rs:24]: (Pedantic)

Rustdoc doesn't understand javadoc syntax, just markdown.

For returns a common thing to do is have a section called Returns (https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md#module-level-vs-type-level-docs)

Maybe for arguments you can do 

```
# Arguments
* `rule_set_name` The name of the ruleset to get
```

but this comment is arguably unnecessary here.

Comment (niklasdewally) [reply conjure_oxide/src/rule_engine/resolve_rules.rs:24]: See also https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html.

You also should probably use /// on each line of a doc comment, that seems to be convention too

File: crates/conjure_rules/src/rule_set.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
+use crate::get_rules;
+use conjure_core::rule::Rule;
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::sync::OnceLock;
+
+/**
+ * A set of rules with a name, priority, and dependencies.
+ *
+ * # Fields
```
Comment (niklasdewally) [inline crates/conjure_rules/src/rule_set.rs:10]: (Pedantic) Fields of a struct are documented above each field - see https://doc.rust-lang.org/rust-by-example/meta/doc.html

**Review by niklasdewally — APPROVED** [review]
Still LGTM :)

I like the new organisation of the module a lot better, and good job on documenting things thoroughly. 

Just left some comments in placed where you've done javadoc style comments instead of rustdoc.

You can generate the documentation for the workspace using cargo doc --workspace; the html generated from this is in target/docs.

**Pull request merged**